src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -0,0 +1,130 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+
+#include "xpmem_impl.h"
+#include "xpmem_noinline.h"
+#include "build_nodemap.h"
+
+int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag_bits)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int i, my_local_rank = -1, num_local = 0;
+    xpmem_segid_t *xpmem_segids = NULL;
+    int local_rank_0 = -1;
+    MPIDU_shm_seg_t shm_seg;
+    MPIDU_shm_barrier_t *shm_seg_barrier = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_INIT_HOOK);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_INIT_HOOK);
+    MPIR_CHKPMEM_DECL(1);
+
+    /* Try to share entire address space */
+    MPIDI_XPMEM_global.segid = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE,
+                                          MPIDI_XPMEM_PERMIT_VALUE);
+    /* 64-bit segment ID or failure(-1) */
+    MPIR_ERR_CHKANDJUMP(MPIDI_XPMEM_global.segid == -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_make");
+    MPL_DBG_MSG_FMT(MPIR_DBG_PT2PT, VERBOSE, (MPL_DBG_FDEST, "xpmem_init: make segid: 0x%lx\n",
+                                              (uint64_t) MPIDI_XPMEM_global.segid));
+
+    MPIR_NODEMAP_get_local_info(rank, size, MPIDI_global.node_map[0], &num_local,
+                                &my_local_rank, &local_rank_0);
+    MPIDI_XPMEM_global.num_local = num_local;
+    MPIDI_XPMEM_global.local_rank = my_local_rank;
+    MPIDI_XPMEM_global.node_group_ptr = NULL;
+
+    /* Allocate temporary shared buffer to exchange xpmem segids with all
```

Comment:
It sounds reasonable, but my concern is XPMEM should locate under CH4 shm module, and in the future, we may add PIP and perhaps other shmem techniques. If we expose this interface to `MPID_Init`, to me it looks a little bit weird.

Reply:
Let me rephrase, what I meant is that since most of future shmmods will likely need POSIX as prerequisite maybe we should split initialization in `MPIDI_SHM_mpi_pre_init_hook` to initialize POSIX shmmod code and `MPIDI_SHM_mpi_post_init_hook` for all the other shmmods. I am not sure this is the right way of doing this.

Reply:
I like the idea of `MPIDI_SHM_mpi_pre|post_init_hook ` ! Maybe we can define the new hooks in this PR (in a separate commit).

Reply:
@minsii please have a look at https://github.com/pmodels/mpich/pull/3930 
this should solve the issue in a much cleaner way.

Reply:
@gcongiu Thanks for implementing it ! We will use `SHM_allgather` here !
@KaimingOuyang Can you please change the code ? I think #3930 will be merged to master before this PR, because it is much simpler :-)

Reply:
This issue was resolved by using the new `MPIDU_Init_shm_put|get` util routines.